streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/service/impl/AppBuildPipeServiceImpl.java

```diff
@@ -210,13 +211,18 @@ public void onStart(PipeSnapshot snapshot) {
               }
             } else {
               if (!app.getDependencyObject().getJar().isEmpty()) {
+                String localUploads = Workspace.local().APP_UPLOADS();
                 // copy jar to local upload dir
                 for (String jar : app.getDependencyObject().getJar()) {
                   File localJar = new File(WebUtils.getAppTempDir(), jar);
-                  Utils.required(localJar.exists());
-                  String localUploads = Workspace.local().APP_UPLOADS();
-                  String uploadJar = localUploads.concat("/").concat(jar);
-                  checkOrElseUploadJar(FsOperator.lfs(), localJar, uploadJar, localUploads);
+                  File uploadJar = new File(localUploads, jar);
+                  if (!localJar.exists() && !uploadJar.exists()) {
+                    throw new ApiAlertException("Missing file: " + jar + ", please upload again");
+                  }
+                  if (localJar.exists()) {
+                    checkOrElseUploadJar(
+                        FsOperator.lfs(), localJar, uploadJar.getAbsolutePath(), localUploads);
+                  }
```

Comment:
Do we need to deal with `!localJar.exists() && uploadJar.exists()` ? This may happen when upgrading.

Reply:
> Do we need to deal with `!localJar.exists() && uploadJar.exists()` ? This may happen when upgrading.

`localJar` is find jar under `$APP_HOME/temp`,
`Uploadjar` is find jars in `$workspace/upload`, `$workspace/upload` is a safe directory. The files uploaded by the user in history will be recorded and will not be deleted with the version upgrade.